# Exploration 2
### 1. 손글씨 분류하기
### 2. 와인 분류하기
### 3. 유방암 여부 진단하기
---
#### scikit-learn 의 내장 예제 데이터셋을 이용하여 위 세가지 모델에 대한 정확도를 예측해보기

In [1]:
import pandas as pd
import numpy

#### 프로젝트를 진행하는데 필요한 공통 라이브러리 import
- 이번 Exploration Node 에서는 공통적으로 pandas와 numpy 라이브러리를 이용하기 때문에 각각의 문제에 들어가기 전에 `import` 해준다.
- pandas는 2차원 배열 데이터를 다루는데 효과적인 도구이며, pandas 라디브러리의 `DataFrame`을 이용해 주어진 데이터셋을 표로 만들어 분석할 수 있다.

## 1. 손글씨 분류

In [2]:
from sklearn.datasets import load_digits

scikit-learn 라이브러리의 예제 데이터셋인 손글씨 데이터 셋을 사용하기 위해 해당 라이브러리로부터 `import` 해준다. &nbsp; 이는 와인, 유방암 진단 에서도 마찬가지!

---

In [3]:
digits = load_digits()
digits_data = digits.data
print(digits.keys())
print(digits_data.shape)

dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])
(1797, 64)


`digits` 변수에 손글씨 데이터셋을 불러와 저장해주고, keys 메서드와 data 메서드를 이용해 담고있는 정보가 무엇인지, 개수는 총 몇개인지를 확인할 수 있다.

---

In [4]:
print(digits.feature_names)
print(digits.target_names)

['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']
[0 1 2 3 4 5 6 7 8 9]


In [5]:
digits_df = pd.DataFrame(data = digits_data, columns = digits.feature_names)
digits_df

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0           0.0        0.0        5.0       13.0        9.0        1.0   
1           0.0        0.0        0.0       12.0       13.0        5.0   
2           0.0        0.0        0.0        4.0       15.0       12.0   
3           0.0        0.0        7.0       15.0       13.0        1.0   
4           0.0        0.0        0.0        1.0       11.0        0.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0           0.0        0.0        0.0        0.0  ...        0.0        0.0   
1           0.0        0.0        0.0        0.0  ...        0.0        0.0   
2           0.0        0.0        0.0        0.0  ...        5.0        0.0   
3           0.0        0.0        0.0        8.0  ...        9.0        0.0   
4           0.0        0.0        0.0        0.0  ...        0.0        0.0   
...         ...        ...        ...        ...  ...        ...        ...   
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0           0.0        0.0        6.0       13.0       10.0        0.0   
1           0.0        0.0        0.0       11.0       16.0       10.0   
2           0.0        0.0        0.0        3.0       11.0       16.0   
3           0.0        0.0        7.0       13.0       13.0        9.0   
4           0.0        0.0        0.0        2.0       16.0        4.0   
...         ...        ...        ...        ...        ...        ...   
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
0           0.0        0.0  
1           0.0        0.0  
2           9.0        0.0  
3           0.0        0.0  
4           0.0        0.0  
...         ...        ...  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[1797 rows x 64 columns]

`DataFrame`을 이용해서 손글씨 데이터를 자료형으로 변환할 수 있다.<br>
Row에는 손글씨 데이터, Column에는 특징들을 넣어주어 1797 $\times$ 64  크기의 매트릭스가 생성된 것을 확인할 수 있다.

---

프로젝트를 절차에 맞게 코드를 다시 적어보면 아래와 같다

In [6]:

# (1) 필요한 모듈 import
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report




# (2) 데이터 준비
digits = load_digits()
digits_data = digits.data
digits_label = digits.target

In [7]:
# (3) train, test 데이터 분리
x_train, x_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

scikit=learn에서 제공하는 `train_test_split`는 주어진 데이터를 학습 데이터와 테스트 데이터로 나눌 수 있는 함수를 말한다. <br>여기서 `test_size = 0.2` 는 학습 데이터와 테스트 데이터의 비를 $8:2$ 로 하여 학습 및 테스트를 진행하겠다는 뜻이다.

### 지도학습과 비지도 학습
#### 학습을 진행하기에 앞서 머신러닝의 학습에 대해 알아보자. <br>
<br>
머신러닝은 크게 *지도학습*(Supervised Learning)과, 비지도학습(Unsupervised Learning)의 두가지로 구분이 되는데, <br>
지도학습은 정답이 있는, 즉 사용자가 분류하고자 하는 라벨링이 명확하게 있는 문제에 대한 학습을 가리킨다. 이번 프로젝트 역시 분류하고자 하는 숫자가 명확하게 존재하므로 지도학습이라고 할 수 있다. <br><br>
반면 비지도학습은 정답이 없는 문제를 학습하는 것으로, 인공지능이 데이터들 간의 공통점을 찾아 스스로 분류하는 학습을 의미한다.

### 지도학습과 비지도 학습
#### 학습을 진행하기에 앞서 머신러닝의 학습에 대해 알아보자. 
---



머신러닝은 크게 **지도학습(Supervised Learning)**과, **비지도학습(Unsupervised Learning)**의 두가지로 구분이 된다.
<br><br>


**지도학습**은 정답이 있는, 즉 사용자가 분류하고자 하는 라벨링이 명확하게 있는 문제에 대한 학습을 가리킨다. <br>이번 프로젝트 역시 분류하고자 하는 숫자가 명확하게 존재하므로 지도학습이라고 할 수 있다. <br><br>
반면 **비지도학습**은 정답이 없는 문제를 학습하는 것으로, 인공지능이 데이터들 간의 공통점을 찾아 스스로 분류하는 학습을 의미한다.

### 지도학습의 종류
---
지도학습은 다시 분류(Classification)와 회귀(Regression)으로 나뉘는데, 분류는 입력받은 특정 카테고리 중 하나로 분류하는 문제를 의미하고 <br>
회귀는 입력받은 데이터에 따라 특정 필드의 수치를 맞히는 문제를 말한다.
<br>
이 중, 이번 프로젝트에서 사용되는 지도학습은 __분류__ 이다.

### 분류 알고리즘
#### 프로젝트에 사용되는 알고리즘을 알아보자
---
#### Decision Tree Classifier
    - Decision Tree, 의사결정나무를 이용한 방법으로, 데이터를 분석하여 이들 사이에 존재하는 패턴을 예측 가능한 규칙들의 조합으로 나타내는 방법.
    - 의사결정나무 자체는 분류와 회귀 모두 가능. 즉, 범주나 연속형 수치 모두 예측이 가능하다는 의미.
    - 결정 경계가 데이터에 수직이라 특정 데이터에만 잘 작동할 가능성이 존재.


#### Random Forest Classifier
    - Random Forest는 Decision Tree 모델 여러개를 합쳐 Decision Tree의 단점을 보완한 모델.
    - Decision Tree를 만드는 데 있어 쓰이는 요소를 무작위로 선정하여 여러개의 Decision Tree를 생성.
    

#### SVM Classifier
    - Support Vector와 Hyperplane을 이용하여 분류를 수행하게 되는 선형 분류 알고리즘.
    - 2차원 공간에서 2개의 클래스만 존재할 때, 이 둘을 구별할 수 있는 선을 그어 구별하는 방법.
    - 이를, 모든 클래스와 구별하게하여 일대다 방법으로 사용해서 분류할 수 있음.

#### SGD Classifier
    - 배치크기가 1인 경사하강법 알고리즘. 
    - 반복이 충분하면 효과는 있지만 노이즈가 심한 단점이 있음.

#### Logistic Regression
    - 가장 널리 알려진 선형 분류 알고리즘으로, 소프트맥스 함수를 사용하여 다중 클래스 분류를 함.
    - 이름은 Regression(회귀) 이지만 실제로 수행하는 알고리즘은 Classification(분류)

### Decision Tree Classifier

In [8]:
# DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier

# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(x_train, y_train)
y_pred = decision_tree.predict(x_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.81      0.81      0.81        42
           2       0.79      0.82      0.80        40
           3       0.79      0.91      0.85        34
           4       0.83      0.95      0.89        37
           5       0.90      0.96      0.93        28
           6       0.84      0.93      0.88        28
           7       0.96      0.82      0.89        33
           8       0.88      0.65      0.75        43
           9       0.78      0.78      0.78        32

    accuracy                           0.86       360
   macro avg       0.86      0.86      0.86       360
weighted avg       0.86      0.86      0.85       360



### Random Forest Classifier

In [9]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=2)
random_forest.fit(x_train, y_train)
y_pred = random_forest.predict(x_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.95      1.00      0.98        42
           2       1.00      0.97      0.99        40
           3       0.97      1.00      0.99        34
           4       0.93      1.00      0.96        37
           5       0.90      1.00      0.95        28
           6       1.00      0.96      0.98        28
           7       0.94      0.97      0.96        33
           8       0.97      0.86      0.91        43
           9       1.00      0.94      0.97        32

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



### SVM Classifier

In [10]:
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(x_train, y_train)
y_pred = svm_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      1.00      0.98        42
           2       1.00      1.00      1.00        40
           3       1.00      1.00      1.00        34
           4       1.00      1.00      1.00        37
           5       0.93      1.00      0.97        28
           6       1.00      1.00      1.00        28
           7       1.00      1.00      1.00        33
           8       1.00      0.93      0.96        43
           9       1.00      0.97      0.98        32

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360



### SGD Classifier

In [11]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_pred = sgd_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        43
           1       0.83      0.95      0.89        42
           2       0.95      1.00      0.98        40
           3       0.94      0.91      0.93        34
           4       0.95      0.95      0.95        37
           5       0.93      1.00      0.97        28
           6       0.93      0.93      0.93        28
           7       0.97      0.97      0.97        33
           8       0.93      0.86      0.89        43
           9       0.96      0.81      0.88        32

    accuracy                           0.94       360
   macro avg       0.94      0.94      0.94       360
weighted avg       0.94      0.94      0.94       360



### Logistic Regression

In [12]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(x_train, y_train)
y_pred = logistic_model.predict(x_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/Users/parkminkyun/miniforge3/envs/tf25/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### 모델의 정확도

|Model|Accuracy|
|:---:|:---:|
|Decision Tree Classifier|0.86|
|Random Forest Classifier|0.97|
|SVM Classifier|0.99|
|SGD Classifier|0.94|
|Logistic Regression|0.95|
    

___
___아래의 와인분류, 유방암 진단 역시 같은 알고리즘에 데이터 셋만 달라진 것이므로 코드에 대한 부가적인 설명인 생략한다.___

## 2. 와인분류

In [13]:

# (1) 필요한 모듈 import
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

# (2) 데이터 준비
wine = load_wine()
wine_data = wine.data
wine_label = wine.target

wine_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [14]:
# (3) train, test 데이터 분리
x_train, x_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.3, 
                                                    random_state=7)

### Decision Tree Classifier

In [51]:
from sklearn.tree import DecisionTreeClassifier

#
decision_tree = DecisionTreeClassifier(random_state = 32)
decision_tree.fit(x_train, y_train)
y_pred = decision_tree.predict(x_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114



### Randon Forest Classifier

In [50]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=23)
random_forest.fit(x_train, y_train)
y_pred = random_forest.predict(x_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        40
           1       0.97      1.00      0.99        74

    accuracy                           0.98       114
   macro avg       0.99      0.97      0.98       114
weighted avg       0.98      0.98      0.98       114



### SVM Classifier

In [46]:
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(x_train, y_train)
y_pred = svm_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114



### SGD Classifier

In [41]:
from sklearn.linear_model import SGDClassifier

sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_pred = sgd_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.65      0.79        40
           1       0.84      1.00      0.91        74

    accuracy                           0.88       114
   macro avg       0.92      0.82      0.85       114
weighted avg       0.90      0.88      0.87       114



### Logistic Regression

In [19]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(x_train, y_train)
y_pred = logistic_model.predict(x_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.80      0.92      0.86        13
           1       0.95      0.88      0.91        24
           2       1.00      1.00      1.00        17

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.92        54
weighted avg       0.93      0.93      0.93        54



/Users/parkminkyun/miniforge3/envs/tf25/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### 모델의 정확도

|Model|Accuracy|
|:---:|:---:|
|Decision Tree Classifier|0.91|
|Random Forest Classifier|0.98|
|SVM Classifier|0.90|
|SGD Classifier|0.88|
|Logistic Regression|0.93|
    

## 3. 유방암 여부 진단

In [20]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report




# (2) 데이터 준비
breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data
breast_cancer_label = breast_cancer.target

breast_cancer_df = pd.DataFrame(data = breast_cancer_data, columns = breast_cancer.feature_names)
breast_cancer_df

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [21]:
# (3) train, test 데이터 분리
x_train, x_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

### Decision Tree Classifier

In [59]:
# DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier

# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(x_train, y_train)
y_pred = decision_tree.predict(x_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.82      0.87        40
           1       0.91      0.96      0.93        74

    accuracy                           0.91       114
   macro avg       0.91      0.89      0.90       114
weighted avg       0.91      0.91      0.91       114



### Random Forest Classifier

In [63]:
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=34)
random_forest.fit(x_train, y_train)
y_pred = random_forest.predict(x_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.90      0.95        40
           1       0.95      1.00      0.97        74

    accuracy                           0.96       114
   macro avg       0.97      0.95      0.96       114
weighted avg       0.97      0.96      0.96       114



### SVM Classifier

In [60]:
from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(x_train, y_train)
y_pred = svm_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.72      0.84        40
           1       0.87      1.00      0.93        74

    accuracy                           0.90       114
   macro avg       0.94      0.86      0.89       114
weighted avg       0.92      0.90      0.90       114



### SGD Classifier

In [55]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_pred = sgd_model.predict(x_test)

print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.81      0.85      0.83        40
           1       0.92      0.89      0.90        74

    accuracy                           0.88       114
   macro avg       0.86      0.87      0.87       114
weighted avg       0.88      0.88      0.88       114



### Logistic Regression

In [26]:
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

logistic_model.fit(x_train, y_train)
y_pred = logistic_model.predict(x_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.85      0.92        40
           1       0.93      1.00      0.96        74

    accuracy                           0.95       114
   macro avg       0.96      0.93      0.94       114
weighted avg       0.95      0.95      0.95       114



/Users/parkminkyun/miniforge3/envs/tf25/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


### 모델의 정확도

|Model|Accuracy|
|:---:|:---:|
|Decision Tree Classifier|0.91|
|Random Forest Classifier|0.96|
|SVM Classifier|0.90|
|SGD Classifier|0.88|
|Logistic Regression|0.95|
    

# 회고
---

## 1. 프로젝트에 사용된 모델에 대한 평가

#### 세가지 프로젝트에 대한 각각의 정확도는 아래와 같다.
|Model|Digits|Wine|Cancer|
|---|---|---|---|
|Decision Tree Classifier|0.86|0.91|0.91|
|Random Forest Classifier|0.97|0.98|0.96|
|SVM Classifier|0.99|0.90|0.90|
|SGD Classifier|0.94|0.88|0.88|
|Logistic Regression|0.95|0.93|0.95|
<br>

위 표를 참고하면, 같은 모델이라도 당연히 데이터셋에 따라 정확도의 차이가 존재하며 Model들 간에도 데이터셋에따라 정확도의 차이가 클 때도 있고, 작을 때도 있는 것을 볼 수 있다. 또한 A라는 데이터셋 에서는 정확도가 더 높았던 모델이 B라는 데이터 셋 에서는 정확도가 다른 모델들에 비해 떨어질 수도 있다는 것을 볼 수 있다.

단, 모델을 평가하는데 있어 정확도로만 판단하기에는 문제가 있을 수 있다. 예를들어 데이터셋이 불균형할 때, 경우에 따라 정확도에 큰 영향을 미쳐 실제 모델의 성능과는 다른 결과가 나올 가능성도 존재한다. 여러가지 클래스 중에 특정 클래스에 대하여 다른 클래스에 비해 정답률이 높을 가능성이 있을 수 있다는 의미이다.

또한 때에 따라 모델이 얼마나 정답을 맞췄느냐 보다 __모델이 얼마나 오답을 내지 않느냐__가 더 중요할 수도 있다. 손글씨 분류와 와인분류의 경우 클래스를 구별하는 것의 오답률과 정답률의 중요도가 비슷할 수 있지만 유방암진단 프로젝트의 경우에 암에 걸린 사람을 걸렸다고 진단하는 것 보다, 암에 걸린 사람을 걸리지 않았다고 판단하는 것이 더 중요할 것이다. <br><br>
이에 대한 성능을 비교하기 위하여 우리는 Precision(정밀도), Recall Sensitivity(재현율), f1 Score 등을 지표로 사용한다. 이 지표들은 모두 오차행렬의 True Positive, True Negative, False Positive, False Negative 을 이용한 지표들이다. <br>

$ Precision = \frac{TP} {TP+FP} $ <br><br>
$ Recall = \frac{TP} {TP+FN} $ <br><br>
$ f1 Score = 2\times\frac{(Precision)(Recall)} {Precision+Recall} $ <br><br>
$ Accuracy = \frac{TP+TN} {TP+TN+FP+FN} $ <br><br>

즉, 손글씨 분류와 와인분류의 경우에는 Accuracy 가, 유방암 진단의 경우에는 Recall 이 더 중요하다고 할 수 있다.

## 3. Reference
[1] https://ratsgo.github.io/machine%20learning/2017/03/26/tree/<br>
[2] https://manisha-sirsat.blogspot.com/2019/04/confusion-matrix.html